### PR TITLE
Add tags

### PR DESCRIPTION
**Pull request by vvscode**
Related to #4

### CODE REVIEW

File: src/exporter.js [hunk 1]
```diff
@@ -124,6 +125,10 @@ export default class {
   }
 }
 
+export const SPACE_REPLACER = String.fromCharCode(160);
```
Comment (ewnd9) [inline src/exporter.js]: Replace it to underscore.

- The original Ruby gem [replaces](https://github.com/albertzak/anki2/blob/5eb49e82309457cf24e1469535784c5ca900c639/lib/anki2.rb#L98) it to underscore characters.

- http://ankisrs.net/docs/manual.html uses same convention 
image

I suppose it is quite possible that Anki clients have logic to transform `underscore_case` to `Human Case` on searching / displaying tags.

File: src/exporter.js [hunk 2]
```diff
@@ -66,8 +66,9 @@ export default class {
     this.media.push({ filename, data });
   }
 
-  addCard(front, back) {
+  addCard(front, back, options = {}) {
```
Comment (ewnd9) [inline src/exporter.js]:
```js
addCard(front, back, { tags = [] } = {}) {
```

File: test/exporter.js [hunk 3]
```diff
@@ -63,3 +63,33 @@ test('Exporter.addCard', t => {
...
+  t.is(exporterUpdateSpy.callCount, 2, 'should made two requests');
+
+  t.is(exporterUpdateSpy.args[ 0 ][ 0 ], `insert into notes values(:id,:guid,:mid,:mod,:usn,:tags,:flds,:sfld,:csum,:flags,:data)`);
+  const notesUpdate = exporterUpdateSpy.args[ 0 ][ 1 ];
+  const notesTags = notesUpdate[ ':tags' ].split(' ');
+  t.is(notesUpdate[ ':sfld' ], front);
+  t.is(notesUpdate[ ':flds' ], front + separator + back);
+  t.is(notesUpdate[ ':mid' ], topModelId);
+
+  t.is(notesTags.length, tags.length);
```
Comment (ewnd9) [inline test/exporter.js]: Just compare that with `deepEqual`. 

```js
t.is(notesTags[2].replace(/\W/g, '_'), tags[2].replace(/\W/g, '_'))
```

Are you testing that `String#replace` is a pure function and returns similar results for similar input and similar arguments?

Comment (vvscode) [inline test/exporter.js]: Yep. String#replace replace only first entry. And at this case I should compare two different strings `some tag` vs `some\u160tag` - so deep Equal don't to what I want

File: src/exporter.js (hunk 1 above)
Comment (vvscode) [inline src/exporter.js]: yep. But then anki displays two tags `vocab` and `check_with_tutor`. I check it before. It's variant, and I agree to replace `\u160` with `_` but it looks ugly

Comment (ewnd9) [inline src/exporter.js]: > But then anki displays two tags vocab and check_with_tutor. I check it before

Thanks for clarifying.

I think I have a better idea: 

```js
':tags': typeof tags === 'string' ? tags : tagsToStr(tags) // tagsToStr replaces to '_'
```

So the end user of the module will have a choice to use a custom separators

File: test/exporter.js (hunk 3 above)
Comment (vvscode) [inline test/exporter.js]: Ok. I move `tagsToStr` into class method to allow override it ( if I as user want's to use `\u160` as before ) and default implementation use `_`

File: src/exporter.js [hunk 4]
```diff
@@ -66,17 +66,24 @@ export default class {
...
 
-  addCard(front, back) {
+  addCard(front, back, { tags } = {}) {
     const { topDeckId, topModelId, separator } = this;
     const note_id = rand();
 
+    let strTags = '';
+    if (typeof tags === 'string'){
+      strTags = tags;
+    } else if (tags instanceof Array) {
```
Comment (ewnd9) [inline src/exporter.js]: Use `Array.isArray` https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray

Comment (vvscode) [inline src/exporter.js]: thx

File: test/exporter.js [hunk 5]
```diff
@@ -63,3 +63,57 @@ test('Exporter.addCard', t => {
...
+  t.is(notesTags[0], tags[0]);
+  t.is(notesTags[1], tags[1]);
+  t.is(notesTags[2].replace(/\W/g, '_'), tags[2].replace(/\W/g, '_'), 'Not the same but similar tag for multiwords');
```
Comment (ewnd9) [inline test/exporter.js]:
```js
t.is(notesTags[2], tags[2].replace(/\W/g, '_'))
```

https://github.com/ewnd9/anki-apkg-export/pull/9/files#r91860519

File: test/exporter.js [hunk 6]
```diff
@@ -83,20 +83,11 @@ test('Exporter.addCard with options (tags is array)', t => {
...
-  t.is(notesTags.length, tags.length);
-  t.is(notesTags[0], tags[0]);
-  t.is(notesTags[1], tags[1]);
-  t.is(notesTags[2].replace(/\W/g, '_'), tags[2].replace(/\W/g, '_'), 'Not the same but similar tag for multiwords');
-
-  t.is(exporterUpdateSpy.args[ 1 ][ 0 ], `insert into cards values(:id,:nid,:did,:ord,:mod,:usn,:type,:queue,:due,:ivl,:factor,:reps,:lapses,:left,:odue,:odid,:flags,:data)`);
-  const cardsUpdate = exporterUpdateSpy.args[ 1 ][ 1 ];
-  t.is(cardsUpdate[ ':did' ], topDeckId);
-  t.is(cardsUpdate[ ':nid' ], notesUpdate[ ':id' ], 'should link both tables via the same note_id');
+  t.deepEqual(notesTags, tags.map(i => i.replace(/ /g, '_')));
```
Comment (ewnd9) [inline test/exporter.js]: `tags.map(tag => `

Variable `i` should be used only as index

Comment (ewnd9) [inline test/exporter.js]: Sorry, missed that in other places. Should be replaced everywhere

File: src/exporter.js [hunk 7]
```diff
@@ -122,6 +129,10 @@ export default class {
   _getFirstVal(query) {
     return JSON.parse(this.db.exec(query)[ 0 ].values[ 0 ]);
   }
+
+  _tagsToStr(tags=[]){
+    return tags.map(i => i.replace(/ /g, '_')).join(' ');
```
Comment (ewnd9) [inline src/exporter.js]: `tags.map(tag`

File: test/exporter.js [hunk 8]
```diff
@@ -63,3 +63,48 @@ test('Exporter.addCard', t => {
...
+  t.is(exporterUpdateSpy.callCount, 2, 'should made two requests');
+
+  t.is(exporterUpdateSpy.args[ 0 ][ 0 ], `insert into notes values(:id,:guid,:mid,:mod,:usn,:tags,:flds,:sfld,:csum,:flags,:data)`);
+  const notesUpdate = exporterUpdateSpy.args[ 0 ][ 1 ];
+  const notesTags = notesUpdate[ ':tags' ].split(' ');
+  t.is(notesUpdate[ ':sfld' ], front);
+  t.is(notesUpdate[ ':flds' ], front + separator + back);
+  t.is(notesUpdate[ ':mid' ], topModelId);
+
+  t.deepEqual(notesTags, tags.map(i => i.replace(/ /g, '_')));
```
Comment (ewnd9) [inline test/exporter.js]: `tags.map(tag`

File: test/test.js [hunk 9]
```diff
@@ -64,3 +64,34 @@ test('check internal structure', async t => {
...
+  const results = await pify(db.all.bind(db))(
+    `SELECT
+      notes.sfld as front,
+      notes.flds as back,
+      notes.tags as tags
+      from cards JOIN notes where cards.nid = notes.id ORDER BY front`);
+  db.close();
+
+  t.deepEqual(results, [
+    { front: front1, back: `${front1}${SEPARATOR}${back1}`, tags: tags1.map(i => i.replace(/ /g, '_')).join(' ') },
```
Comment (ewnd9) [inline test/test.js]: `tags.map(tag`

**Comment by ewnd9** [comment]
Great work :tada: Released as a new major version `v3.0.0`

**Pull request merged**